### DISCUSSION TITLE
Cannot complete content using minibuffer-complete in xref-find-definitions

### DISCUSSION THREAD

**Discussion by BooAA** [joaotavora/eglot#1192] [Q&A]
I'm using the built-in tab completion system. The project I'm working on is Linux kernel and the language server used is clangd. Whenever using `xref-find-definitions`, after I type something to try to filter out the candidates, I cannot use tab (`minibuffer-complete`) to complete the common part. Even I filter the number of candidates down to only one, pressing `tab` still display the `*completion*` buffer and no completion happen.

 I have noticed there is a `(eglot-indirection-joy  (styles external))` in `completion-category-default`, is the behavior described above caused by this settings? I have tried to set `completion-category-defaults` to `nil`, but the value will be restored whenever a new completion is triggered. Has anyone encountered this before? How should I make the tab completion work just as usual?

Screenshot: filter to only one candidate, but  still no completion trigger. 
![Screenshot_2023-03-27_00-57-59](https://user-images.githubusercontent.com/32809182/227791700-54a74ce7-1629-450d-bf5c-012919f36a5d.png)

Also I found it strange that I cannot just enter the exact name in the minibuffer and press return to go to definition. The only way is to use `M-v` to jump to the `*completion*` buffer and select items there. e.g. 

![Screenshot_2023-03-27_01-03-28](https://user-images.githubusercontent.com/32809182/227791958-eeacc1cb-6404-48a1-9501-9606ce74e683.png)

**Comment by joaotavora** [comment]
I've read your report three times now and I cannot understand exactly what is happening.  Please try to report his according to https://joaotavora.github.io/eglot/#Troubleshooting-Eglot

You can skip step 3, as I have `clangd` installed.  In step 4, if this only happens with the linux kernel as the project, please show where you obtained it and, importantly attach a `compile_commands.json` file.  If this happens with smaller projects, try to show one of those smaller projects.  

Focus on a very clear step 6.

**Reply by BooAA** [reply]
I come up a very simple test case, hope that explain things clear.

Test file:

```c
void my_test_func() {
	return;
}

int main() {
	return 0;
}
```

Problem 1:

1. `M-x` eglot, and select `clangd`
2. `C-u M-.`, and type `my_test`
3. press `tab`

**What I expected**: complete the input to `my_test_func` and display "sole completion" in echo area. No \*completion\* buffer should be displayed. 

**What really happens**: no additional character is inserted for completion, the \*completion\* buffer is shown with just one candidate (screenshot below)
![Screenshot_2023-03-27_01-48-28](https://user-images.githubusercontent.com/32809182/227794355-365140f4-d066-4285-85b2-e461718c0d05.png)

 

Problem 2:
`M-x` eglot, and select `clangd`
2. `C-u M-.`, and type `my_test_func`
3. press `return`

**What I expected**: jump to the definition of my_test_func

**What really happens**: no definitions found for my_test_func (screenshot below)
![Screenshot_2023-03-27_01-49-08](https://user-images.githubusercontent.com/32809182/227794415-e370b7b7-1e7f-4b0a-a729-b512783c4dc6.png)

Hope this helps. Thanks.

**Reply by joaotavora** [reply]
> Hope this helps. Thanks.

It does, I reproduced it. 

This doesn't work as smoothly as it does in other `M-.` scenarios, I'm afraid.  But it's not terrible, especially if you use something like `fido-vertical-mode`.

I'll see what I can do.

**Comment by BooAA** [comment]
> I'll see what I can do.

No worries. You have done so many great jobs to have eglot upstream and bring this awesome LSP package into Emacs. Your work saves my life and makes me more productive. Really appreciate.

> But it's not terrible, especially if you use something like fido-vertical-mode.

I have a hybrid workflow that use the build-in system when I know exactly what I want, and use things like helm/ivy/vertico when I want to fuzzy search for something I don't remember. So having this work normally can free my finger from pressing `M-v` to jump to the \*completoin\* buffer to select the item.

Just out of curiosity, is this related to the `(eglot-indirection-joy  (styles external))`  settings that I mentioned in comment#1? Since I observe that same input can got different completions candidates returned when using different language server. (e.g. clangd vs ccls). If this is caused by using external server to do sorting and filtering, what if we just use emacs' build-in completion-styles so that we can have those features back?  (Ignore my proposal if that's totally wrong, as I know nothing about eglot's internal and the LSP protocol)

**Reply by joaotavora** [reply]
`eglot-indirection-joy` is just.... the indirection.  The "external" completion style is its own package, and could be the reason for this.

And _no_, we _can't_ use Emacs's built-in completions styles :-( The `;;; Commentary:` of `external-completion.el` should explain why.
